src/merkle/partial_mt/mod.rs

```diff
@@ -74,6 +78,80 @@ impl PartialMerkleTree {
+    /// - If the depth is 0 or is greater than 64.
+    /// - The number of entries exceeds the maximum tree capacity, that is 2^{depth}.
+    /// - The provided entries contain an insufficient set of nodes.
+    pub fn with_leaves<R, I>(entries: R) -> Result<Self, MerkleError>
+    where
+        R: IntoIterator<IntoIter = I>,
+        I: Iterator<Item = (NodeIndex, RpoDigest)> + ExactSizeIterator,
+    {
+        // a map where the key is the depth of the node, and the value is the vector of indexes and
+        // hashes of nodes with this depth
+        let mut layer_nodes: BTreeMap<u8, Vec<(u64, RpoDigest)>> = BTreeMap::new();
+
+        let mut leaves = BTreeSet::new();
+        let mut nodes = BTreeMap::new();
+
+        for (node_index, hash) in entries.into_iter() {
+            leaves.insert(node_index);
+            nodes.insert(node_index, hash);
+            layer_nodes
+                .entry(node_index.depth())
+                .and_modify(|layer_vec| layer_vec.push((node_index.value(), hash)))
+                .or_insert(vec![(node_index.value(), hash)]);
+        }
+
+        // check if the number of leaves can be accommodated by the tree's depth; we use a min
+        // depth of 63 because we consider passing in a vector of size 2^64 infeasible.
+        let max = (1_u64 << 63) as usize;
+        if layer_nodes.len() > max {
+            return Err(MerkleError::InvalidNumEntries(max, layer_nodes.len()));
+        }
+
+        // Get maximum depth from leaves
+        let max_depth = *layer_nodes.last_key_value().unwrap_or((&0, &vec![])).0;
+
+        // Calculate and add internal nodes
+        // The idea is to calculate parent nodes for nodes in a certain layer, starting from the
+        // deepest one
+        for depth in (1..max_depth + 1).rev() {
+            let layer = layer_nodes.get(&depth).cloned().unwrap_or(vec![]);
+            for (index_value, hash) in layer.iter() {
+                // create NodeIndex from depth and index_value
+                let index = NodeIndex::new(depth, *index_value)?;
+                // create parent node index
+                let mut parent_index = index;
+                parent_index.move_up();
+
+                // checks that we don't push the same parent node to the layer nodes
+                if !nodes.contains_key(&parent_index) {
+                    let sibling_hash = nodes
+                        .get(&index.sibling())
+                        .ok_or(MerkleError::NodeNotInSet(index.sibling()))?;
+                    let parent_hash = Rpo256::merge(&index.build_node(*hash, *sibling_hash));
+
+                    layer_nodes
+                        .entry(depth - 1)
+                        .and_modify(|layer_vec| layer_vec.push((parent_index.value(), parent_hash)))
+                        .or_insert(vec![(parent_index.value(), parent_hash)]);
+                    nodes.insert(parent_index, parent_hash);
+                }
+            }
+        }
```

Comment:
We should avoid extra cloning and insertions here. If we switch `node_layers` to `BTreeMap<u8, Vec<u64>>` we could do something like this:

```Rust
for (depth, layer) in layers.into_iter().rev() {
    for index_value in layer {
        let index = NodeIndex::new(depth, index_value)?;
        let node = nodes.get(&index).ok_or(...);
        let sibling = nodes.get(&index.sibling()).ok_or(...);
        // some checks here
        let parent = Rpo256::merge(&index.build_node(*node, *sibling));
        nodes.insert(index.move_up(), parent);
    }
}
```

Reply:
I'm not sure we can avoid insertions into the `layers` map here because we still need to understand which nodes have been added to the current layer on the previous cycle to continue the loop. Without that, for example, in the case where all leaves are placed in only one layer (like in regular `MerkleTree`) we will calculate the layer above it and stop. Is there an option to mutate a map while iterating on it without `unsafe` code or some `RefCell` stuff, without cloning the layer?

Reply:
Ah - I see! Yes, this complicates things. It may be possible to keep a single `parents` vector which gets updated with each iteration - but I haven't really thought through this.

Reply:
If I get the idea with `parents` vector right, in that case the meaning of using a layers map is lost is lost. But most likely I didn't get the idea.

Reply:
Would something like this work?
```Rust
let mut layer_iter = layers.into_values().rev();
let mut current_layer = layer_iter.next().unwrap();
let mut parent_layer = layer_iter.next().unwrap();

for depth in (1..max_depth + 1).rev() {
    for index_value in current_layer {
        // get the parent node index
        let parent_node = NodeIndex::new(depth - 1, index_value / 2)?;

        // Check if the parent hash was already calculated. In about half of the cases, we
        // don't need to do anything.
        if !nodes.contains_key(&parent_node) {
            // create current node index
            let index = NodeIndex::new(depth, index_value)?;

            // get hash of the current node
            let node = nodes.get(&index).ok_or(MerkleError::NodeNotInSet(index))?;
            // get hash of the sibling node
            let sibling = nodes
                .get(&index.sibling())
                .ok_or(MerkleError::NodeNotInSet(index.sibling()))?;
            // get parent hash
            let parent = Rpo256::merge(&index.build_node(*node, *sibling));

            // add index value of the calculated node to the parents layer
            parent_layer.push(parent_node.value());
            // add index and hash to the nodes map
            nodes.insert(parent_node, parent);
        }
    }

    // set current_layer = parent_layer and parent_layer = layer_iter.next()
    current_layer = layer_iter.next().unwrap();
    core::mem::swap(&mut current_layer, &mut parent_layer);
}
```
You will need to pre-populate `layers` with all layers between 0 and `max_depth` - but that should be pretty simple.